// kijiji-scraper/dist/index.js
"use strict";
// index.ts
/* Exports the kijiji-scraper submodules */
Object.defineProperty(exports, "__esModule", { value: true });
exports.locations = exports.categories = exports.ScraperType = exports.search = exports.AdInfo = exports.Ad = void 0;
var ad_1 = require("./lib/ad");
Object.defineProperty(exports, "Ad", { enumerable: true, get: function () { return ad_1.Ad; } });
var scraper_1 = require("./lib/scraper");
Object.defineProperty(exports, "AdInfo", { enumerable: true, get: function () { return scraper_1.AdInfo; } });
var search_1 = require("./lib/search");
Object.defineProperty(exports, "search", { enumerable: true, get: function () { return search_1.search; } });
var helpers_1 = require("./lib/helpers");
Object.defineProperty(exports, "ScraperType", { enumerable: true, get: function () { return helpers_1.ScraperType; } });
var categories_1 = require("./lib/categories");
Object.defineProperty(exports, "categories", { enumerable: true, get: function () { return categories_1.categories; } });
var locations_1 = require("./lib/locations");
Object.defineProperty(exports, "locations", { enumerable: true, get: function () { return locations_1.locations; } });
